lib/schedulers.go

```diff
@@ -0,0 +1,331 @@
+	sortedConfigs := scs.GetSortedSchedulerConfigs()
+
+	// Combine the steps and requirements from all different schedulers, and
+	// sort them by their time offset, counting the schedulers' startTimes as
+	// well.
+	type trackedStep struct {
+		ExecutionStep
+		configID int
+	}
+	trackedSteps := []trackedStep{}
+	for configID, config := range sortedConfigs { // orderly iteration over a slice
+		configStartTime := config.GetStartTime()
+		configSteps := config.GetExecutionRequirements(executionSegment)
+		for _, cs := range configSteps {
+			cs.TimeOffset += configStartTime // add the scheduler start time to the step time offset
+			trackedSteps = append(trackedSteps, trackedStep{cs, configID})
+		}
+	}
+	// Sort by (time offset, config id). It's important that we use stable
+	// sorting algorithm, since there are could be steps with the same time from
+	// the same scheduler and their order is important.
+	sort.SliceStable(trackedSteps, func(a, b int) bool {
+		switch {
+		case trackedSteps[a].TimeOffset < trackedSteps[b].TimeOffset:
+			return true
+		case trackedSteps[a].TimeOffset == trackedSteps[b].TimeOffset:
+			return trackedSteps[a].configID < trackedSteps[b].configID
+		default:
+			return false
+		}
+	})
+
+	// Go through all of the sorted steps from all of the schedulers, and
+	// build a new list of execution steps that consolidates all of their
+	// requirements. If multiple schedulers have an execution step at exactly
+	// the same time offset, they will be combined into a single new execution
+	// step with the sum of the values from the previous ones.
+	currentTimeOffset := time.Duration(0)
+	currentPlannedVUs := make([]uint64, len(scs))
+	currentMaxUnplannedVUs := make([]uint64, len(scs))
+	sum := func(data []uint64) (result uint64) { // sigh...
+		for _, val := range data {
+			result += val
+		}
+		return result
+	}
+	consolidatedSteps := []ExecutionStep{}
+	addCurrentStepIfDifferent := func() {
+		newPlannedVUs := sum(currentPlannedVUs)
+		newMaxUnplannedVUs := sum(currentMaxUnplannedVUs)
+		stepsLen := len(consolidatedSteps)
+		if stepsLen == 0 ||
+			consolidatedSteps[stepsLen-1].PlannedVUs != newPlannedVUs ||
+			consolidatedSteps[stepsLen-1].MaxUnplannedVUs != newMaxUnplannedVUs {
+
+			consolidatedSteps = append(consolidatedSteps, ExecutionStep{
+				TimeOffset:      currentTimeOffset,
+				PlannedVUs:      newPlannedVUs,
+				MaxUnplannedVUs: newMaxUnplannedVUs,
+			})
+		}
+	}
+	for _, step := range trackedSteps {
+		// If the time offset is different, create a new step with the current values
+		if step.TimeOffset != currentTimeOffset {
+			addCurrentStepIfDifferent()
+			currentTimeOffset = step.TimeOffset
+		}
+		currentPlannedVUs[step.configID] = step.PlannedVUs
+		currentMaxUnplannedVUs[step.configID] = step.MaxUnplannedVUs
+	}
+	addCurrentStepIfDifferent() // Add the last step
+	return consolidatedSteps
```

Comment:
Can we copy the slice at this point so we don't waste space .. although a small amount. 
Also maybe preallocating the slice ?

Reply:
I'm confused - how can we pre-allocate the slice when we don't know its eventual size? And what's the wasted space? Are you concerned that the doubling of the slice `cap` by `append()`? If that's the case, then I'd say it's not worth it, after all, `GetFullExecutionRequirements()` is basically used only in once place in the code...

Reply:
ping @MStoykov :arrow_up:

Reply:
In this particular case, we can't actually calculate it... Or we can - that's what we do in the lines above, but we can only do it iteratively... And again, there's absolutely no point in `copy()`-ing the resulting slice, even ignoring the overhead of needlessly copying stuff, since we'd just save a tiny bit of memory from something that's called only once in the execution of a k6 script - basically the definition of premature optimization. Resolving this, since this PR is huge enough to not need more bikeshedding.